lib/s3/line_stream.ex

```diff
@@ -0,0 +1,47 @@
+defmodule S3.LineStream do
+  @moduledoc """
+  Provide an S3 object as a line-by-line stream.
+  """
+
+  @doc """
+  Returns a stream of lines for a text file in an S3 bucket. Each element of
+  the stream will be a single line instead of a binary chunk.
+  """
+  @spec lines(Enumerable.t()) :: Enumerable.t()
+  def lines(binary_stream) do
+    binary_stream
+    |> Stream.chunk_while("", &split_chunk/2, &last_line/1)
+    |> Stream.flat_map(& &1)
+  end
```

Comment:
This might be simpler with `Stream.transform/3` instead? https://hexdocs.pm/elixir/Stream.html#transform/3

It takes an accumulator and the current record, and lets you return another Enumerable (in this case a list of lines) and the new accumulator (the remainder of the current line).

Reply:
This is actually how I had it implemented first. The problem that I ran into and couldn't figure out a solution to—and quite likely I missed something—is that the last line was getting dropped and not making it into the results. When I got it working this way (which isn't much different in terms of complexity), it seemed good enough to go with, but I'm very open to going back to transform as well.

Reply:
In that case, you probably want `Steam.transform/4` which gives you an `after_fn` called when the stream ends: https://hexdocs.pm/elixir/Stream.html#transform/4

Reply:
I looked into using the `after_fn`. Based on experience and some forum posts it seemed like the takeaway was that anything returned from this function is not included in the Stream output (and that the documentation needs improvement…) I added [a ticket](https://app.asana.com/0/1199957631529739/1200094157954322) to remind myself to come back to this. It sure seems like `Stream.transform` *should* work, and I also think this `binary_stream` might have potential to be a useful standalone package.